mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -252,79 +259,221 @@ impl NexusChild {
             return Err(ChildError::ChildReadOnly {});
         }
 
-        let bdev = self.bdev.as_ref();
-        let desc = self.bdev_handle.as_ref();
-
-        if bdev.is_none() || desc.is_none() {
-            return Err(ChildError::ChildInvalid {});
-        }
+        let bdev = match self.bdev.as_ref() {
+            Some(bdev) => bdev,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
-        let bdev = bdev.unwrap();
-        let desc = desc.unwrap();
+        let desc = match self.bdev_handle.as_ref() {
+            Some(desc) => desc,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
         let block_size = bdev.block_len();
 
-        let primary = u64::from(block_size);
-        let secondary = bdev.num_blocks() - 1;
-
+        //
+        // Protective MBR
         let mut buf = desc
             .dma_malloc(block_size as usize)
             .context(LabelAlloc {})?;
-
-        self.read_at(primary, &mut buf)
-            .await
-            .context(LabelRead {})?;
-
-        let mut label = GPTHeader::from_slice(buf.as_slice());
-        if label.is_err() {
-            warn!(
-                "{}: {}: The primary label is invalid!",
-                self.parent, self.name
-            );
-            self.read_at(secondary, &mut buf)
-                .await
-                .context(LabelRead {})?;
-            label = GPTHeader::from_slice(buf.as_slice());
+        let mbr = self.read_mbr(&mut buf).await?;
+
+        let mut header: Vec<GPTHeader> = Vec::with_capacity(2);
+        let active: GPTHeader;
+
+        //
+        // GPT header(s)
+        match self.read_header(u64::from(block_size), &mut buf).await {
+            Ok(primary) => {
+                active = primary;
+                header.push(primary);
+                match self
+                    .read_header(
+                        (bdev.num_blocks() - 1) * u64::from(block_size),
+                        &mut buf,
+                    )
+                    .await
+                {
+                    Ok(secondary) => {
+                        header.push(secondary);
+                        self.check_consistency(&primary, &secondary)?;
+                    }
+                    Err(_) => {
```

Comment:
should we capture the error and log it as part of the warn!?

Reply:
done